lib/AST/DiagnosticEngine.cpp

```diff
@@ -964,11 +966,21 @@ void DiagnosticEngine::emitDiagnostic(const Diagnostic &diagnostic) {
 }
 
 const char *DiagnosticEngine::diagnosticStringFor(const DiagID id,
-                                                  bool printDiagnosticName) {
+                                                  bool printDiagnosticName,
+                                                  bool fromWarningsAsErrors) {
+  const char *diagString = diagnosticStrings[(unsigned)id];
+  if (!printDiagnosticName && !fromWarningsAsErrors)
+    return diagString;
+
+  SmallString<128> buffer(diagString);
+  if (fromWarningsAsErrors)
+    buffer.append(" [-warnings-as-errors]");
   if (printDiagnosticName) {
-    return debugDiagnosticStrings[(unsigned)id];
+    buffer.append(" [");
+    buffer.append(diagnosticStringIDs[(unsigned) id]);
+    buffer.append("]");
   }
-  return diagnosticStrings[(unsigned)id];
+  return TransactionStrings.insert(buffer.str()).first->getKeyData();
```

Comment:
This string looks like it'll be leaked if the current diagnostic isn't part of a transaction. We don't always have a top level transaction and right now `TransactionStrings` only gets cleared when the top level transaction commits or aborts. I suppose we could also clear it at the end of `emitDiagnostic` if `TransactionCount` is 0, but that feels like a hack to me

Reply:
Any suggestions for where the string should be allocated?

Reply:
It might not be a bad idea to inline `diagnosticStringFor` into `diagnosticInfoForDiagnostic`. `diagnosticInfoForDiagnostic` can then use a `StringSet` passed in from `emitDiagnostic`. At the end of `emitDiagnostic` all of the strings can be thrown out.